File: tests/rpc/create-member.ts
import * as anchor from "@project-serum/anchor";
import { SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID,ASSOCIATED_TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { Ctx, Member } from "../ctx/ctx";

export async function createMemberRPC(ctx: Ctx, member: Member) {
    await ctx.program.methods.createMember()
    .accounts({
        factory: ctx.PDAS.factory.key,
        stakeTokenMint: ctx.PDAS.factory.stakeTokenMint,
        beneficiary: member.beneficiary.publicKey,
        member: member.key,
        vaultFree: member.vaultFree,
        systemProgram: SystemProgram.programId,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        rent: anchor.web3.SYSVAR_RENT_PUBKEY,
    })
    .signers([member.beneficiary])
    .rpc();
}

File: programs/staking/src/context/create_member.rs
use anchor_lang::prelude::*;
use crate::account::*;
use anchor_spl::token::{TokenAccount, Token, Mint};
use anchor_spl::associated_token::AssociatedToken;

#[derive(Accounts)]
pub struct CreateMember<'info> {
    #[account(
        seeds = [Factory::PDA_SEED],
        bump = factory.bump,
        has_one = stake_token_mint,
    )]
    pub factory: Account<'info, Factory>,
    pub stake_token_mint: Box<Account<'info, Mint>>,
    #[account(
        init,
        payer = beneficiary,
        space = 8 + Member::SPACE,
        seeds = [
            beneficiary.to_account_info().key.as_ref(),
            factory.to_account_info().key.as_ref(),
        ],
        bump,
    )]
    pub member: Account<'info, Member>,
    #[account(
        init,
        payer = beneficiary,
        associated_token::mint = stake_token_mint,
        associated_token::authority = member,
    )]
    pub vault_free: Box<Account<'info, TokenAccount>>,
    #[account(mut)]
    pub beneficiary: Signer<'info>,
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub rent: Sysvar<'info, Rent>,
}
